File: src/main/java/com/moovie/backend/data/remote/tmdb/model/TmdbResponses.java
package com.moovie.backend.data.remote.tmdb.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;
import java.util.Map;

public class TmdbResponses {

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbMovieResponse(
            int id,
            String title,
            String overview,
            @JsonProperty("poster_path") String posterPath,
            @JsonProperty("backdrop_path") String backdropPath,
            @JsonProperty("vote_average") double voteAverage,
            @JsonProperty("release_date") String releaseDate
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbMovieListResponse(
            int page,
            @JsonProperty("total_pages") int totalPages,
            @JsonProperty("total_results") int totalResults,
            List<TmdbMovieResponse> results
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbMovieDetailResponse(
            int id,
            String title,
            String overview,
            @JsonProperty("poster_path") String posterPath,
            @JsonProperty("backdrop_path") String backdropPath,
            @JsonProperty("vote_average") double voteAverage,
            @JsonProperty("release_date") String releaseDate,
            String tagline,
            Integer runtime,
            List<TmdbGenre> genres,
            TmdbCredits credits,
            @JsonProperty("watch/providers") TmdbWatchProvidersWrapper watchProviders,
            TmdbMovieListResponse similar,
            TmdbReviewListResponse reviews
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbCredits(
            List<TmdbCrewMember> crew,
            List<TmdbCastMember> cast
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbCrewMember(
            String name,
            String job
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbCastMember(
            String name,
            String character
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbWatchProvidersWrapper(
            Map<String, TmdbWatchProviderRegion> results
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbWatchProviderRegion(
            List<TmdbWatchProviderEntry> flatrate
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbWatchProviderEntry(
            @JsonProperty("provider_name") String providerName,
            @JsonProperty("logo_path") String logoPath
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbReviewListResponse(
            int page,
            @JsonProperty("total_pages") int totalPages,
            @JsonProperty("total_results") int totalResults,
            List<TmdbReview> results
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbReview(
            String id,
            String author,
            String content,
            @JsonProperty("created_at") String createdAt,
            @JsonProperty("author_details") TmdbAuthorDetails authorDetails
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbAuthorDetails(
            Double rating
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbGenre(
            int id,
            String name
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbGenreListResponse(
            List<TmdbGenre> genres
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbCountry(
            @JsonProperty("iso_3166_1") String iso,
            @JsonProperty("english_name") String englishName
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TmdbLanguage(
            @JsonProperty("iso_639_1") String iso,
            @JsonProperty("english_name") String englishName
    ) {
    }
}

File: src/main/kotlin/org/example/data/remote/model/TmdbResponses.kt
package org.example.data.remote.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class TmdbMovieResponse(
    val id: Int,
    val title: String? = null,
    val overview: String? = null,
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("vote_average") val voteAverage: Double? = null,
    @SerialName("release_date") val releaseDate: String? = null
)

@Serializable
data class TmdbMovieListResponse(
    val page: Int = 1,
    @SerialName("total_pages") val totalPages: Int = 1,
    @SerialName("total_results") val totalResults: Int = 0,
    val results: List<TmdbMovieResponse> = emptyList()
)

@Serializable
data class TmdbMovieDetailResponse(
    val id: Int,
    val title: String? = null,
    val overview: String? = null,
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("vote_average") val voteAverage: Double? = null,
    @SerialName("release_date") val releaseDate: String? = null,
    val tagline: String? = null,
    val runtime: Int? = null,
    val genres: List<TmdbGenre> = emptyList(),
    val credits: TmdbCredits? = null,
    @SerialName("watch/providers") val watchProviders: TmdbWatchProvidersWrapper? = null,
    val similar: TmdbMovieListResponse? = null,
    val reviews: TmdbReviewListResponse? = null
)

@Serializable
data class TmdbCredits(
    val crew: List<TmdbCrewMember> = emptyList(),
    val cast: List<TmdbCastMember> = emptyList()
)

@Serializable
data class TmdbCrewMember(
    val name: String? = null,
    val job: String? = null
)

@Serializable
data class TmdbCastMember(
    val name: String? = null,
    val character: String? = null
)

@Serializable
data class TmdbWatchProvidersWrapper(
    val results: Map<String, TmdbWatchProviderRegion> = emptyMap()
)

@Serializable
data class TmdbWatchProviderRegion(
    val flatrate: List<TmdbWatchProviderEntry> = emptyList()
)

@Serializable
data class TmdbWatchProviderEntry(
    @SerialName("provider_name") val providerName: String? = null,
    @SerialName("logo_path") val logoPath: String? = null
)

@Serializable
data class TmdbReviewListResponse(
    val page: Int = 1,
    @SerialName("total_pages") val totalPages: Int = 1,
    @SerialName("total_results") val totalResults: Int = 0,
    val results: List<TmdbReview> = emptyList()
)

@Serializable
data class TmdbReview(
    val id: String? = null,
    val author: String? = null,
    val content: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("author_details") val authorDetails: TmdbAuthorDetails? = null
)

@Serializable
data class TmdbAuthorDetails(
    val rating: Double? = null
)

@Serializable
data class TmdbGenre(
    val id: Int,
    val name: String? = null
)

@Serializable
data class TmdbGenreListResponse(
    val genres: List<TmdbGenre> = emptyList()
)

@Serializable
data class TmdbCountry(
    @SerialName("iso_3166_1") val iso: String,
    @SerialName("english_name") val englishName: String? = null
)

@Serializable
data class TmdbLanguage(
    @SerialName("iso_639_1") val iso: String,
    @SerialName("english_name") val englishName: String? = null
)
